extensions/cornerstone-dicom-sr/src/getSopClassHandlerModule.ts

```diff
@@ -205,69 +210,101 @@ async function _load(displaySet, servicesManager: AppTypes.ServicesManager, exte
+ * @param servicesManager - The services manager.
+ */
 function _checkIfCanAddMeasurementsToDisplaySet(
   srDisplaySet,
   newDisplaySet,
   dataSource,
   servicesManager: AppTypes.ServicesManager
 ) {
   const { customizationService } = servicesManager.services;
+
   let unloadedMeasurements = srDisplaySet.measurements.filter(
     measurement => measurement.loaded === false
   );
 
+  /** All measurements were loaded */
   if (unloadedMeasurements.length === 0) {
-    // All already loaded!
     return;
   }
 
-  if ((!newDisplaySet) instanceof ImageSet) {
-    // This also filters out _this_ displaySet, as it is not an ImageSet.
+  if (!(newDisplaySet instanceof ImageSet)) {
     return;
   }
 
   if (newDisplaySet.unsupported) {
     return;
   }
 
-  const { sopClassUids } = newDisplaySet;
+  const { sopClassUids, images } = newDisplaySet;
 
-  // Check if any have the newDisplaySet is the correct SOPClass.
+  /** Check correct SOPClassUID */
   unloadedMeasurements = unloadedMeasurements.filter(measurement =>
-    measurement.coords.some(coord =>
-      sopClassUids.includes(coord.ReferencedSOPSequence.ReferencedSOPClassUID)
-    )
+    measurement.coords.some(coord => {
+      if (coord.ReferencedSOPSequence === undefined && coord.ReferencedFrameOfReferenceSequence) {
+        for (let i = 0; i < images.length; ++i) {
+          const imageMetadata = images[i];
+          if (imageMetadata.FrameOfReferenceUID !== coord.ReferencedFrameOfReferenceSequence) {
+            continue;
+          }
+
+          const sliceNormal = [0, 0, 0];
+          const orientation = imageMetadata.ImageOrientationPatient;
+          sliceNormal[0] = orientation[1] * orientation[5] - orientation[2] * orientation[4];
+          sliceNormal[1] = orientation[2] * orientation[3] - orientation[0] * orientation[5];
+          sliceNormal[2] = orientation[0] * orientation[4] - orientation[1] * orientation[3];
+
+          let distanceAlongNormal = 0;
+          for (let j = 0; j < 3; ++j) {
+            distanceAlongNormal += sliceNormal[j] * imageMetadata.ImagePositionPatient[j];
+          }
+
+          /** Assuming 2 mm tolerance */
+          if (Math.abs(distanceAlongNormal - coord.GraphicData[2]) > 2) {
+            continue;
+          }
+
+          coord.ReferencedSOPSequence = {
+            ReferencedSOPClassUID: imageMetadata.SOPClassUID,
+            ReferencedSOPInstanceUID: imageMetadata.SOPInstanceUID,
+          };
+
+          break;
+        }
+      }
+
+      return (
+        coord.ReferencedSOPSequence &&
+        sopClassUids.includes(coord.ReferencedSOPSequence.ReferencedSOPClassUID)
+      );
+    })
   );
 
   if (unloadedMeasurements.length === 0) {
-    // New displaySet isn't the correct SOPClass, so can't contain the referenced images.
+    /** New displaySet isn't the correct SOPClassso can't contain the referenced images */
     return;
   }
 
   const SOPInstanceUIDs = [];
```

Comment:
Could you change this to a Set<string> as that turns the algorithm from O(n^2) to O(n) because it doesn't need to check every instance for equality.  It should be only a few lines changed.

Reply:
Updated!